src/raftstore/store/peer.rs

```diff
@@ -119,30 +124,55 @@ impl ReadIndexQueue {
     }
 
     /// update the read index of the requests that before the specified id.
-    fn advance(&mut self, id: &[u8], read_index: u64) {
-        if let Some(i) = self.reads.iter().position(|x| x.binary_id() == id) {
-            for pos in 0..=i {
-                let req = &mut self.reads[pos];
-                let index = req.read_index.get_or_insert(read_index);
-                if *index > read_index {
-                    *index = read_index;
+    fn advance_replica_reads<T: IntoIterator<Item = (Uuid, u64)>>(&mut self, states: T) {
+        let (mut min_changed_offset, mut max_changed_offset) = (usize::MAX, 0);
+        for (uuid, index) in states {
+            if let Some(offset) = self.contexts.remove(&uuid) {
+                let offset = offset.checked_sub(self.handled_cnt).unwrap();
+                self.ready_cnt = cmp::max(self.ready_cnt, offset + 1);
+
+                if let Some(occur_index) = self.reads[offset].read_index {
+                    if occur_index < index {
+                        continue;
+                    }
                 }
+                self.reads[offset].read_index = Some(index);
+                min_changed_offset = cmp::min(min_changed_offset, offset);
+                max_changed_offset = cmp::max(max_changed_offset, offset);
+                continue;
             }
-            if self.ready_cnt < i + 1 {
-                self.ready_cnt = i + 1;
-            }
-        } else {
-            error!(
+            debug!(
                 "cannot find corresponding read from pending reads";
-                "id"=>?id, "read-index" =>read_index,
+                "uuid" => ?uuid, "read-index" => index,
             );
         }
+
+        if self.ready_cnt > 1 && max_changed_offset > 0 {
+            self.fold(min_changed_offset, max_changed_offset);
+        }
+    }
+
+    fn fold(&mut self, mut min_changed_offset: usize, max_changed_offset: usize) {
```

Comment:
Do we need `min_changed_offset`?  the purpose of this function is to update all `offset < max_changed_offset` that `read_index < r_idex`.